lib/puma/server.rb

```diff
@@ -598,6 +603,21 @@ def notify_safely(message)
     end
     private :notify_safely
 
+    def set_timeout_for_idle(idle_timeout)
+      return unless idle_timeout
+
+      @idle_timeout_at = Process.clock_gettime(Process::CLOCK_MONOTONIC) + idle_timeout
+    end
+    private :set_timeout_for_idle
+
+    # Number of seconds until the idle timeout elapses.
+    def timeout_for_idle
+      return unless defined?(@idle_timeout_at)
+
+      [@idle_timeout_at - Process.clock_gettime(Process::CLOCK_MONOTONIC), 0].max
```

Comment:
I just feel like checking system time every time we read data is just not going to scale. I don't feel comfortable with this.

Reply:
That's a valid concern, I'll see if I can come up with an alternative.

Reply:
@nateberkopec You mentioned above that this runs in a super hot loop. Do you have a sense for how hot? On my Mac, you can check the monotonic clock roughly 11.5 million times per second in Ruby code. It's definitely faster than the system clock.

<details><summary>Benchmark code</summary>

```
require "benchmark/ips"

Benchmark.ips do |x|
  x.report("monotonic clock") { Process.clock_gettime(Process::CLOCK_MONOTONIC) }
  x.report("system clock") { Time.now }

  x.compare!
end
```

</details>

```
Warming up --------------------------------------
     monotonic clock     1.146M i/100ms
        system clock   333.694k i/100ms
Calculating -------------------------------------
     monotonic clock     11.526M (± 0.9%) i/s -     58.462M in   5.072420s
        system clock      3.303M (± 1.6%) i/s -     16.685M in   5.052375s

Comparison:
     monotonic clock: 11526338.7 i/s
        system clock:  3303211.7 i/s - 3.49x  slower
```

Reply:
Resolved with: 9c4a15a158b554f67e59f33378eee10abeec8c14.

Reply:
Sounds good. I just wanted to understand the concerns around performance here since checking the monotonic clock is surprisingly cheap. It's actually slightly cheaper than `Object.new`:

<details><summary>Benchmark code</summary>

```ruby
require "benchmark/ips"

n = 0
Benchmark.ips do |x|
  x.report("do nothing") {}
  x.report("allocate an object") { Object.new }
  x.report("monotonic clock") { Process.clock_gettime(Process::CLOCK_MONOTONIC) }
  x.report("system clock") { Time.now }

  x.compare!
end
```

</details>

```
Warming up --------------------------------------
          do nothing     2.399M i/100ms
  allocate an object     1.097M i/100ms
     monotonic clock     1.132M i/100ms
        system clock   325.979k i/100ms
Calculating -------------------------------------
          do nothing     23.868M (± 0.9%) i/s -    119.961M in   5.026342s
  allocate an object     10.932M (± 0.9%) i/s -     54.832M in   5.016068s
     monotonic clock     11.405M (± 0.9%) i/s -     57.747M in   5.063726s
        system clock      3.259M (± 1.1%) i/s -     16.299M in   5.001941s

Comparison:
          do nothing: 23868388.0 i/s
     monotonic clock: 11404972.2 i/s - 2.09x  slower
  allocate an object: 10932187.5 i/s - 2.18x  slower
        system clock:  3258906.8 i/s - 7.32x  slower
```